modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientChannel.java

```diff
@@ -141,52 +174,108 @@
         throws ClientConnectionException, ClientAuthorizationException {
+        ClientRequestFuture pendingReq = pendingReqs.get(reqId);
 
-        final int MIN_RES_SIZE = 8 + 4; // minimal response size: long (8 bytes) ID + int (4 bytes) status
+        assert pendingReq != null : "Pending request future not found for request " + reqId;
 
-        int resSize = new BinaryHeapInputStream(read(4)).readInt();
+        try {
+            while (true) {
+                if (rcvLock.tryLock()) {
+                    try {
+                        if (!pendingReq.isDone())
+                            processNextResponse();
+                    }
+                    finally {
+                        rcvLock.unlock();
+                    }
+                }
 
-        if (resSize < 0)
+                try {
+                    byte[] payload = pendingReq.get(PAYLOAD_WAIT_TIMEOUT);
+
+                    if (payload == null || payloadReader == null)
+                        return null;
+
+                    return payloadReader.apply(new PayloadInputStream(this, payload));
+                }
+                catch (IgniteFutureTimeoutCheckedException ignore) {
+                    // Next cycle if timed out.
+                }
+            }
+        }
+        catch (IgniteCheckedException e) {
+            if (e.getCause() instanceof ClientError)
+                throw (ClientError)e.getCause();
+
+            if (e.getCause() instanceof ClientException)
+                throw (ClientException)e.getCause();
+
+            throw new ClientException(e.getMessage(), e);
+        }
+        finally {
+            pendingReqs.remove(reqId);
+        }
+    }
+
+    /**
+     * Process next response from the input stream and complete corresponding future.
+     */
+    private void processNextResponse() throws ClientProtocolError, ClientConnectionException {
+        int resSize = readInt();
+
+        if (resSize <= 0)
             throw new ClientProtocolError(String.format("Invalid response size: %s", resSize));
 
-        if (resSize == 0)
-            return null;
+        long bytesReadOnStartReq = totalBytesRead;
+
+        long resId = readLong();
+
+        ClientRequestFuture pendingReq = pendingReqs.get(resId);
 
-        BinaryInputStream resIn = new BinaryHeapInputStream(read(MIN_RES_SIZE));
+        if (pendingReq == null)
+            throw new ClientProtocolError(String.format("Unexpected response ID [%s]", resId));
 
-        long resId = resIn.readLong();
+        int status = 0;
 
-        if (resId != reqId)
-            throw new ClientProtocolError(String.format("Unexpected response ID [%s], [%s] was expected", resId, reqId));
+        BinaryInputStream resIn;
 
-        int status = resIn.readInt();
+        status = readInt();
 
-        if (status != 0) {
-            resIn = new BinaryHeapInputStream(read(resSize - MIN_RES_SIZE));
+        int hdrSize = (int)(totalBytesRead - bytesReadOnStartReq);
```

Comment:
An approach with counting `totalBytesRead` seems fragile to me because counting methods should be used everywhere. I think counting a header size could be encapsulated into a method or an object counting just bytes while reading header only.

Reply:
In current protocol version header size is fixed, but in next versions, header size depends on different conditions (status, topology change, etc), so to calculate header size we need either to increment the counter each time readLong/readInt/... is called or increment the counter inside readLong/readInt/... There are much more parameters to read from the header in next protocol versions, so incrementing counter inline is more error-prone and more verbose.

Reply:
Good to know it. But it still be error-prone as long as we have a input streams at hand which allow us to read data without counting.

Reply:
Perhaps we can extract input stream and `read...` methods to a new class, but it seems overcomplicated for this task to me.